File: 0101.symmetruc_tree/solution.java
// Definition for a binary tree node.
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

class Solution {
    public boolean isSymmetric(TreeNode root) {
        return isMirror(root, root);
    }
    
    public boolean isMirror(TreeNode left, TreeNode right){
        if (left == null && right == null)
            return true;
        if (left == null || right == null)
            return false;
        return left.val == right.val && isMirror(left.left, right.right) && isMirror(right.left, left.right);
    }
}

File: 0101.symmetruc_tree/solution.py
from utils import TreeNode


class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        return self.is_mirror(root, root)

    def is_mirror(self, left, right):
        if left is None and right is None:
            return True
        if not left or not right:
            return False
        return (
            left.val == right.val and self.is_mirror(left.left, right.right) and self.is_mirror(right.left, left.right)
        )
